### DISCUSSION TITLE
운영진에서 준 BaseLine과 현재 구현 코드의 차이

### DISCUSSION THREAD

**Discussion by intrandom5** [boostcampaitech4nlp2/level1_semantictextsimilarity_nlp-level1-nlp-13#13] [Show and tell]
제가 구현해 놨던 코드와 baseline 코드의 성능 차이가 큰 것을 보고 경악을 금치 못할 수 없었습니다.
도대체 뭐가 문제였던걸까요?

>     
    def tokenizing(self, dataframe):
        data = []
        for idx, item in tqdm(dataframe.iterrows(), desc='tokenizing', total=len(dataframe)):
            # 두 입력 문장을 [SEP] 토큰으로 이어붙여서 전처리합니다.
            text = '[SEP]'.join([item[text_column] for text_column in self.text_columns])
            outputs = self.tokenizer(text, add_special_tokens=True, padding='max_length', truncation=True)
            data.append(outputs['input_ids'])
        return data
위의 코드는 pytorch lightning baseline의 코드입니다. 보면 문장을 따로 제공하는 SBERT가 아니라, 문장을 한 번에 처리하는 BERT 모델의 input을 만들고 있다는 것을 알 수 있습니다.
그러면 저희의 코드도 그렇게 수정을 해봐야겠죠?
>     
    class KorSTSDatasets_for_BERT(KorSTSDatasets):
        def __init__(self, dir, model_name):
            super(KorSTSDatasets_for_BERT, self).__init__(dir, model_name)

        def __getitem__(self, idx):
            data = self.s1[idx][:-1] + [self.sep_id] + self.s2[idx][1:]
            data = torch.IntTensor(data)
            label = float(self.y[idx])
            return data, label
다음은 모델입니다.
>
    def __init__(self, model_name)
        self.plm = transformers.AutoModelForSequenceClassification.from_pretrained(
            pretrained_model_name_or_path=model_name, num_labels=1)
        # Loss 계산을 위해 사용될 L1Loss를 호출합니다.
        self.loss_func = torch.nn.L1Loss()
        def forward(self, x):
            x = self.plm(x)['logits']

            return x
AutoModelForSequenceClassification이라는 모듈을 사용해서 모델을 불러왔군요...
loss function도 MSE가 아니라 MAE를 쓰고 있습니다.
저희 코드에서도 수정해줘야겠지요.
>
    class BERT_base_Model(nn.Module):
        def __init__(self, model_name):
            super(BERT_base_Model, self).__init__()
            self.bert = BertForSequenceClassification.from_pretrained(model_name, num_labels=1)
            self.linear = nn.Linear(self.bert.config.hidden_size, 1)
            self.similarity = nn.CosineSimilarity(dim=-1)
        def forward(self, src_ids):
            outputs = self.bert(src_ids)

            return outputs.logits
그 외에도 learning rate(=1e-5), batch_size(=64)로 수정을 해준 뒤에 다시 학습을 해 봤습니다.
그리고 baseline model을 참고해서 pearson 상관계수 점수도 logging하도록 코드를 수정했습니다.

빠른 결과를 얻기 위해 base model은 klue/bert-base를 사용했습니다. 결과는...
![image](https://user-images.githubusercontent.com/68216991/198175479-b4d1a7d0-91b5-4ea9-a96b-daf77b056d29.png)
우와.. loss도 원래 실험들보다 훨씬 작고 pearson 점수도 0.86정도? 나왔습니다. 괜찮은 성능을 보이는군요.
이렇게 저희의 코드 자체에는 문제가 없다는 것을 확인할 수 있었습니다. 다만 모델을 어떻게 구현하냐에 따라서 성능이 차이가 크군요.
이 모델 부분은 추후 확인해서 다시 정리할 필요가 있을 것 같습니다.
괜찮은 성능의 모델을 얻은 김에 결과 분석도 살짝 해봅시다.
우선 예측 결과와 라벨을 비교해 보죠.
![scatter_result](https://user-images.githubusercontent.com/68216991/198176406-1c930cf9-9b5a-4763-b38c-27c79b994478.png)
이것만으로는 얼마나 잘 맞췄다라고 확인하기는 힘들지만.. 기존처럼 특정 값에 예측값이 몰리고 있지는 않다는 것을 확인할 수 있습니다.

모델이 혹시 많이 틀리는 원천 source가 있지는 않을까요? 한 번 확인해 봅시다..
>
    x = np.arange(6)
    sources = ["nsmc-sampled", "nsmc-rtt", "slack-sampled", "slack-rtt", "petition-sampled", "petition-rtt"]
    values = [df[df["원천"] == src]["점수 차이"].mean() for src in sources]
    plt.figure(figsize=(15, 5))
    plt.bar(x, values)
    plt.xticks(x, sources)
    plt.show()
![image](https://user-images.githubusercontent.com/68216991/198177250-cee9e2e0-ca2c-424c-a1fb-8d08eeda782f.png)
보면 source에 따라 점수 차이가 0.5점 내외 정도로 source 별 점수 차이는 크지 않은 것을 확인할 수 있습니다... 최소한 source에는 문제가 없는 것 같군요.
이런 식으로 앞으로도 모델 출력 결과와 데이터를 분석하면서 모델의 성능을 올려나가면 좋을 것 같습니다...! 파이팅...
